zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZooKeeperUpdatingListenerBuilder.java

```diff
@@ -0,0 +1,201 @@
+ * invoking {@link #connectTimeout(Duration)}, {@link #connectTimeoutMillis(long)},
+ * {@link #sessionTimeout(Duration)} or {@link #sessionTimeoutMillis(long)} will raise a
+ * {@link IllegalStateException}.
+ *
+ * <pre>{@code
+ * ZooKeeperUpdatingListener listener =
+ *     new ZooKeeperUpdatingListenerBuilder(curatorFramework, "/myProductionEndpoints")
+ *         .nodeValueCodec(NodeValueCodec.DEFAULT)
+ *         .build();
+ * ServerBuilder sb = new ServerBuilder();
+ * sb.addListener(listener);
+ * }</pre>
+ * */
+public final class ZooKeeperUpdatingListenerBuilder {
+    private CuratorFramework client;
+    private final String connectionStr;
+    private final String zNodePath;
+    private int connectTimeoutMillis = ZooKeeperDefaults.DEFAULT_CONNECT_TIMEOUT_MS;
+    private int sessionTimeoutMillis = ZooKeeperDefaults.DEFAULT_SESSION_TIMEOUT_MS;
+    private Endpoint endpoint;
+    private NodeValueCodec nodeValueCodec = NodeValueCodec.DEFAULT;
+
+    /**
+     * Creates a {@link ZooKeeperUpdatingListenerBuilder} with a {@link CuratorFramework} instance and a zNode
+     * path.
+     *
+     * @param client the curator framework instance
+     * @param zNodePath the ZooKeeper node to register
+     */
+    public ZooKeeperUpdatingListenerBuilder(CuratorFramework client, String zNodePath) {
+        this.client = requireNonNull(client, "client");
+        this.connectionStr = null;
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        checkArgument(!this.zNodePath.isEmpty(), "zNodePath can't be empty");
+    }
+
+    /**
+     * Creates a {@link ZooKeeperUpdatingListenerBuilder} with a ZooKeeper connection string and a zNode path.
+     *
+     * @param connectionStr the ZooKeeper connection string
+     * @param zNodePath the ZooKeeper node to register
+     */
+    public ZooKeeperUpdatingListenerBuilder(String connectionStr, String zNodePath) {
+        this.connectionStr = requireNonNull(connectionStr, "connectionStr");
+        checkArgument(!this.connectionStr.isEmpty(), "connectionStr can't be empty");
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        checkArgument(!this.zNodePath.isEmpty(), "zNodePath can't be empty");
+    }
+
+    private void ensureInternalClient() {
+        checkState(client == null,
+                   "This method is allowed only when created with a connection string.");
+    }
+
+    /**
+     * Sets the connect timeout.
+     *
+     * @param connectTimeout the connect timeout
+     *
+     * @throws IllegalStateException if this builder is constructed with
+     *                               {@link #ZooKeeperUpdatingListenerBuilder(CuratorFramework, String)}
+     */
+    public ZooKeeperUpdatingListenerBuilder connectTimeout(Duration connectTimeout) {
+        requireNonNull(connectTimeout, "connectTimeout");
+        checkArgument(!connectTimeout.isZero() && !connectTimeout.isNegative(),
```

Comment:
How about removing this line?

Reply:
That would result in incorrect exception message. `sessionTimeout: ...` vs `sessionTimeoutMillis: ...`

Reply:
Got it! please forget my comments!